File: implement-trie-prefix-tree.py
from typing import Optional, Self

class TrieNode:
    def __init__(self):
        self.children = {}
        self.end = False

    def get_or_create(self, char: str) -> Self:
        if char in self.children:
            return self.children[char]
        node = TrieNode()
        self.children[char] = node
        return node
    
    def get(self, char: str) -> Optional[Self]:
        return self.children.get(char, None)

class Trie:

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for ch in word:
            node = node.get_or_create(ch)
        node.end = True

    def search(self, word: str) -> bool:
        node = self.root
        for ch in word:
            node = node.get(ch) 
            if node is None:
                return False
        return node.end

    def startsWith(self, prefix: str) -> bool:
        node = self.root
        for ch in prefix:
            node = node.get(ch) 
            if node is None:
                return False
        return True

File: implement-trie-prefix-tree.cpp
using namespace std;

#define Index(ch) (ch - 'a')

struct TrieNode {
	TrieNode *children[26];
	bool isEnd;

	TrieNode() {
		isEnd = false;
		for (auto &c: children) c = nullptr;
	}

	TrieNode *get_or_create(char ch) {
		int i = Index(ch);
		if (children[i] != nullptr) {
			return children[i];
		}
		TrieNode *node = new TrieNode();
		children[i] = node;
		return node;
	}

	TrieNode *get(char ch) {
		return children[Index(ch)];
	}
};

class Trie {
	TrieNode *root;
public:
    Trie() {
		root = new TrieNode();
    }

    void insert(string word) {
        TrieNode *node = root;
		for (char ch: word) {
			node = node->get_or_create(ch);
		}
		node->isEnd = true;
    }

    bool search(string word) {
        TrieNode *node = root;
		for (char ch: word) {
			node = node->get(ch);
			if (node == nullptr) return false;
		}
		return node->isEnd;
    }

    bool startsWith(string prefix) {
        TrieNode *node = root;
		for (char ch: prefix) {
			node = node->get(ch);
			if (node == nullptr) return false;
		}
		return true;
    }
};
